File: solutions/csharp/nth-prime/1/NthPrime.cs
// Honouring the divine guidance of Shree DR.MDD

using System;
using System.Collections.Generic;
using System.Linq;

public static class NthPrime
{
    public static int Prime(int rank)
    {
        if (rank == 0)
        {
            throw new ArgumentOutOfRangeException();
        }

        int candidate = 1;
        int found = 0;

        while (found < rank)
        {
            candidate++;
            int div = 2;
            while (div <= candidate)
            {
                if (candidate % div == 0)
                {
                    break;
                }
                div++;
            }
            if (div == candidate)
            {
                found++;
            }
        }

        return candidate;
    }
}

File: solutions/zig/nth-prime/1/nth_prime.zig
// For Shree DR.MDD

const std = @import("std");
const mem = std.mem;

pub const PrimeError = error{
    InvalidInput,
};

fn checkPrime(x: usize) bool {
    if (x <= 1) return false;
    if (x <= 3) return true;
    if (x % 2 == 0 or x % 3 == 0) return false;

    var i: usize = 5;
    while (i * i <= x) : (i += 6) {
        if (x % i == 0 or x % (i + 2) == 0) return false;
    }

    return true;
}

pub fn prime(allocator: mem.Allocator, target: usize) !usize {
    _ = allocator;
    if (target == 0) return PrimeError.InvalidInput;

    var found: usize = 0;
    var candidate: usize = 1;

    while (found < target) : (candidate += 1) {
        if (checkPrime(candidate)) {
            found += 1;
        }
    }

    return candidate - 1;
}
